Fixed #26226 -- Made related managers honor the queryset used for prefetching their results.

charettes opened the pull request:
@loic I'd like to have your feedback about this slightly backward incompatible change.

If I remember correctly such an issue was discussed during the implementation of the `Prefetch` object.

charettes commented on tests/prefetch_related/tests.py:
```diff
@@ -695,6 +695,14 @@ def test_custom_qs(self):
+        self.assertEqual(
+            list(person.houses.all()),
+            list(person.houses.all().all()),
+        )
```
The fact this assertion fails without the fix shows how the POLA is not respected here.

loic commented:
I don't recall we discussed this issue specifically, it was more along the lines that if you filter things out, `.all()` becomes ambiguous, hence with filtering it's better to use `to_attr` than to cache the result in the manager. I like the fix though, I'll comment inline.

loic commented on django/db/models/query.py:
```diff
@@ -1606,8 +1609,11 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
-                qs = getattr(obj, to_attr).all()
+                manager = getattr(obj, to_attr)
+                if leaf and lookup.queryset is not None:
+                    qs = manager._filter_queryset(lookup.queryset)
```
I don't see fixing `model.relation.all().all()` as much of a backwards incompatibility, but the introduction of `_filter_queryset` is more troublesome. I haven't checked but I suspect projects that do fishy stuff with relation managers like taggit or mptt may hit an `AttributeError`. I can imagine 3 options:
1. Add `hasattr` to the if clause.
2. Leave it as it is and document the incompatibility, custom related managers are more or less private APIs and it's easy to fix.
3. Issue a deprecation warning when `_filter_queryset` is missing then switch to 2. after the deprecation.

I'm fine with both 2. and 3., 1. doesn't give a chance to projects to fix this inconsistency in their code.

Btw the relation manager of `GenericRelation` probably needs a similar treatment.

charettes commented:
I adjusted the `GenericRelation`'s manager and added a test.

Both 2. and 3. are fine with me. I would favor 3. as it isn't much effort anyway.

What do you think of the `_filter_queryset()` wording? Do you think `_bind_queryset()` would make more sense?

loic commented:
How about doing away with the extra method and having `get_queryset(self, ignore_cache=False)`?

loic commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -502,23 +502,26 @@ def __call__(self, **kwargs):
+            db = self._db or router.db_for_read(self.model, instance=self.instance)
+            empty_strings_as_null = connections[db].features.interprets_empty_strings_as_nulls
+            queryset._add_hints(instance=self.instance)
+            if self._db:
```
It's orthogonal to this PR but this `self._db` in place of just `db` looks wrong.

charettes replied:
yeah I guess we could revisit this in another PR.

loic commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -502,23 +502,26 @@ def __call__(self, **kwargs):
+            empty_strings_as_null = connections[db].features.interprets_empty_strings_as_nulls
+            queryset._add_hints(instance=self.instance)
+            if self._db:
+                queryset = queryset.using(self._db)
```
ditto.

loic commented:
As discussed briefly on IRC, we could just use a clone of `rel_qs` which will contain all the relevant filters. I believe `getattr(obj, to_attr).all()` was only used to get a cache holder. PR https://github.com/django/django/pull/6167 shows this approach.

charettes commented:
As pointed out on #6167 this won't work as soon as queryset with multiple results is used with prefetch_related. The previous tests only happened to work because they were only testing against a single record.

charettes commented:
> How about doing away with the extra method and having `get_queryset(self, ignore_cache=False)`?

I'm not sure we want to hijack the documented `get_queryset()` method to add an undocumented kwarg.

charettes commented:
@loic I used `rel_qs` instead of `manager.all()`.

loic commented on django/db/models/query.py:
```diff
@@ -1606,8 +1609,11 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
+                if leaf and lookup.queryset is not None:
+                    qs = manager._filter_queryset(lookup.queryset)
+                else:
+                    qs = manager._filter_queryset(rel_qs)
```
By using `rel_qs` we can simplify to: `qs = getattr(obj, to_attr)._filter_queryset(rel_qs)` and do away with the `if/else`. What do you think?

charettes replied:
Right!

charettes commented:
By the way, weren't you're `.all()` overrides already broken by the actual `manager.all()` usages here?

loic commented:
Never hit the problem so I guess I never prefetched on those models, but yeah they would. Retrospectively that was a gratuitous usage of `.all()`, could have been a `get_queryset()` all along.

loic commented:
Re `_filter_queryset()` vs `_bind_queryset()` I like both, you pick! Since we are at it, I'll toss another one to the mix: `_apply_rel_filters()` :P

charettes commented:
`_apply_rel_filters()` sounds less generic to me, will go with that.

charettes commented:
Added the deprecation warnings for related managers with no `_apply_rel_filters(queryset)` method.

charettes commented:
@loic I fell back to use `lookup.queryset` if available and `manager.get_queryset()` (formerly `manager.all()`) instead of `rel_qs` after all.

Since we have all the required details to generate a queryset without double filtering (`field IN (...) AND field = ...`) I think we should favor this option.

charettes commented:
Adjusted and merged the `setUpData()` commit into master c30086159d6b6e33a10f18ec0d11719636bc1a94.

timgraham commented on docs/internals/deprecation.txt:
```diff
@@ -141,6 +141,9 @@ details on these changes.
   ``Meta.default_related_name`` is set will be removed.
 
+* The shim for supporting custom related manager classes with no
+  ``_apply_rel_filters`` method will be removed.
```
without a
filters()

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -732,6 +732,17 @@ use the default_related_name ``bars``::
 
     >>> Foo.object.get(bars=bar)
 
+Custom related manager must define a ``_apply_rel_filters()`` method
```
managers

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -732,6 +732,17 @@ use the default_related_name ``bars``::
+it defines a ``_apply_rel_filters()`` method.
+
+This method must accept a :class:`~django.db.models.query.QuerySet` instance
+as its single argument and return a version appropriately filetered for the
```
a filtered version of the queryset...
(is this a private API such that we don't need to document it elsewhere?)

charettes replied:
Yes it's a private API but some third-party apps out there such as django-taggit do some funky stuff with related managers and Loïc mentioned it could be worth noting the change in the release notes.

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -732,6 +732,17 @@ use the default_related_name ``bars``::
+--------------------------------------------------------------------
+
+If you defined a custom manager class available through
+:meth:`~django.db.models.query.QuerySet.prefetch_related` you must make sure
```
Custom manager classes available through prefetch_related must define a ...
(not sure if an example is needed or if this is advanced enough that anyone doing this will know what it means?)

charettes replied:
I don't think an example is required, I believe this will affect a really small subset of users.

I'll make sure to create PR on django-taggit's project and search for other project defining custom manager classes.

timgraham commented on tests/prefetch_related/tests.py:
```diff
@@ -704,6 +716,30 @@ def test_nested_prefetch_related_are_not_overwritten(self):
         )
 
+    def test_apply_rel_filters_deprecation_shim(self):
+        # Simulate a missing `_apply_rel_filters` method.
```
some names have backticks, in these comments and some don't. Personally I would just use _apply_rel_filters() but as long as you're consistent I don't mind either way.

timgraham commented:
Some docstrings for `_apply_rel_filters()` might be helpful.

djowett commented:
I know this is an old patch, but this requirement should be documented outside of just the Release Notes.
